array_analyzer/extract/txt_parser.py

```diff
@@ -38,34 +53,124 @@ def create_xml_dict(path_):
+        array_params['bg_offset'] = float(layout['@background_offset'])
+        array_params['bg_thickness'] = float(layout['@background_thickness'])
+        array_params['max_diam'] = float(layout['@max_diameter'])
+        array_params['min_diam'] = float(layout['@min_diameter'])
 
-    # replicates
-    repl = doc['configuration']['well_configurations']['configuration']['array']['spots']['multiplet']
+        # not present in .xml but manually added here:
+        array_params['pixel_size_scienion'] = 0.0049
+        array_params['pixel_size_octopi'] = 0.00185
+    except Exception as ex:
+        raise AttributeError(f"exception while parsing .xml : {ex}")
 
-    params = dict()
-    params['rows'] = int(layout['@rows'])
-    params['columns'] = int(layout['@cols'])
-    params['v_pitch'] = float(layout['@vspace'])
-    params['h_pitch'] = float(layout['@hspace'])
-    params['spot_width'] = float(layout['@expected_diameter'])
-    params['bg_offset'] = float(layout['@background_offset'])
-    params['bg_thickness'] = float(layout['@background_thickness'])
-    params['max_diam'] = float(layout['@max_diameter'])
-    params['min_diam'] = float(layout['@min_diameter'])
+    return fiduc, spots, repl, array_params
 
-    # todo: add param for different imaging conditions (magnificaiton, camera pixel size, etc..)
-    params['pixel_size_scienion'] = 0.0049  # scienion camera
-    params['pixel_size_octopi'] = 0.00185  # octopi camera
 
-    return fiduc, spots, repl, params
+def create_csv_dict(path_):
+    """
+    Looks for three .csv files:
+        "array_parameters.csv" contains array printing parameters as well as hardware parameters
+        "array_format_type.csv" contains fiducial and control names, locations
+        "array_format_antigen.csv" contains specific antigen names for only diagnostic spots
+    Then, parses the .csvs and creates the four dictionaries:
+    :param path_: list
+        of strings to the .csv paths
+    :return:
+    """
+    # assign names to each of the types == params, spot types, antigens
+    fiduc = list()
+    csv_antigens = list()
+    array_params = dict()
+
+    for meta_csv in path_:
+        with open(meta_csv, newline='') as csv_file:
+            csv_reader = csv.reader(csv_file, delimiter=',')
+            for row in csv_reader:
+                # header row
+                if "Parameter" in str(row[0]) or '' in str(row[0]):
+                    continue
+
+                # parse params
+                elif "array_format_parameters" in meta_csv:
+                    array_params[row[0]] = row[1]
+
+                # parse fiducials
+                elif "array_format_type" in meta_csv:
+                    for col, value in enumerate(row[1:]):
+                        pos = {'@row': str(row[0]),
+                               '@col': str(col - 1),
+                               '@spot_type': str(value)}
+                        fiduc.append(pos)
```

Comment:
Is the csv metadata format also in array like xlsx format? Here you convert the metadata into dictionary first then later convert dictionaries into numpy arrays. But if the info is already in array format in the csv you can convert them into numpy arrays directly using Pandas function without doing a two-step conversion

Reply:
The .xml data is not in array format.  From my a quick search, it appears pandas can not read .xmls into dataframes either.  (please ignore the .csv parser, this PR focuses on only .xml and .xlsx)

But the decision with this function (txt_parser.create_xlsx_dict) was to stay consistent with existing machinery.  I do think you're right -- if we simply parse the .xlsx sheets directly into an array, it simplifies that parsing.  There's more to say here but let's leave it for place (a new issue?)